src/work_pool.c

```diff
@@ -274,26 +287,23 @@ work_pool_submit(struct work_pool *pool, struct work_pool_entry *work)
 		/* queue is draining */
 		return (0);
 	}
-	pthread_mutex_lock(&pool->pqh.qmutex);
-
-	if (0 < pool->pqh.qcount--) {
-		struct work_pool_thread *wpt = (struct work_pool_thread *)
-			TAILQ_FIRST(&pool->pqh.qh);
 
-		/* positive for waiting worker(s) */
-		TAILQ_REMOVE(&pool->pqh.qh, &wpt->pqe, q);
-		wpt->work = work;
-
-		/* Note: the mutex is the pool _head,
-		 * but the condition is per worker,
-		 * making the signal efficient!
-		 */
+	pthread_mutex_lock(&pool->pqh.qmutex);
+	/*
+	 * Insert in work queue so that running thread can
+	 * pickup without scheduling.
+	 */
+	TAILQ_INSERT_TAIL(&pool->pqh.qh, &work->pqe, q);
+	struct work_pool_thread *wpt = TAILQ_LAST(&pool->wptqh, work_pool_s);
+	if (wpt) {
+		pool->pqh.qcount--;
+		TAILQ_REMOVE(&pool->wptqh, wpt, wptq);
+		assert(!wpt->wakeup);
+		wpt->wakeup = true;
 		pthread_cond_signal(&wpt->pqcond);
 	} else {
-		/* negative for task(s) */
-		TAILQ_INSERT_TAIL(&pool->pqh.qh, &work->pqe, q);
+		assert(pool->pqh.qcount == 0);
```

Comment:
Essentially, this thread queues the work. If there is a sleeping thread, it wakes it up. This is exactly same as existing code, correct, except that the old code is waking up first thread and you are waking up the last one in the list.  The old is optimizing the thread to pick up work by setting ->work, and in your case, the thread should get the work from the task list. Is this the actual change that gained you perf or something else I am missing here?

Reply:
Yes this is the change which gave boost.
So what's happening now as compared to old one is, the thread I am waking up here may not see any work in the queue and it will again go in wait, as there could be existing running thread which picked up the work in queue without going for wait.
With old code, this was happening only if worker queue is negative for tasks, that means we already hit thread_max.
I am trying to achieve this without hitting thread_max, as it causes lot of scheduling overhead which is significant for low latency workloads like 8k cached reads.
We did some experiments with reducing thread_max to lower value, we got best numbers with thread_max = 8, it gave best results for 8k cached reads but regressed in write workloads as it was IO bound and needed more threads.